lib/phase/flatten.js

```diff
@@ -522,6 +537,75 @@ function convertSteps(jobs) {
     return newJobs;
 }
 
+/**
+ * Get full stage name
+ * @param  {String} stageName               Stage name
+ * @param  {String} type                    Type of stage job (setup or teardown)
+ * @return {String}                         Full stage name (e.g. stage@deploy:setup)
+ */
+function getFullStageName({ stageName, type }) {
+    return `${STAGE_PREFIX}${stageName}:${type}`;
+}
+
+/**
+ * Flatten stage setup and teardown into jobs object
+ * @param  {Object} doc               Document that went through structural parsing
+ */
+function flattenStageSetupAndTeardownJobs(doc) {
+    const newJobs = clone(doc.jobs);
+    const stages = clone(doc.stages);
+
+    if (stages) {
+        Object.keys(stages).forEach(stageName => {
+            const { setup, teardown, jobs: stageJobs } = stages[stageName];
+
+            stageJobs.forEach(jobName => {
+                if (newJobs[jobName]) {
+                    newJobs[jobName].stage = { name: stageName };
+                    // Handle startFrom
+                    if (
+                        newJobs[jobName].requires &&
+                        newJobs[jobName].requires.includes(`~${getFullStageName({ stageName, type: 'setup' })}`)
+                    ) {
+                        stages[stageName].startFrom = stages[stageName].startFrom
```

Comment:
Also `startFrom` is not defined in `stage` domain/DB model. This might be temporary field just utilized in `workflow-parser`. If that is the case, we should keep this localized to `workflow-parser`.

i.e. Instead of `config-parser` checking for the presence of `setup` in `requires` and setting `startFrom` which is then used by `workflow-parser`, we should make `workflow-parser` directly check for the presence of `setup` in `requires`.